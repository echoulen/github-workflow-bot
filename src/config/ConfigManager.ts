import {Config} from "@type/Config";

export class ConfigManager {
  public static getConfig(): Config {
    return {
      integration: process.env.integration as any,
      repository: {
        url: process.env.repoUrl,
        token: process.env.token,
      },
      notify: {
        enable: process.env.enableNotify === "true",
        token: process.env.notifyToken,
      },
      lgtm: {
        enable: process.env.enableLGTM === "true",
        count: Number(process.env.lgtmCount),
      },
    };
  }
}
