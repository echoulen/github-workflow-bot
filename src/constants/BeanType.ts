export enum BeanTypes {
  WEBHOOK = "WEBHOOK",
  API_EXECUTOR = "API_EXECUTOR",
}
