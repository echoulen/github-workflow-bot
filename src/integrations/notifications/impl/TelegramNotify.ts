import {Notify} from "@notifications/Notify";
import {Notification} from "@type/notification/Notification";
import {List} from "immutable";
import {injectable} from "inversify";
import TelegramBot from "node-telegram-bot-api";
import "reflect-metadata";

const regex = new RegExp("/start");

@injectable()
export class TelegramNotify implements Notify {
  private telegramBot: TelegramBot;
  private chatroomIDs: List<number> = List().asMutable();

  public start(token: string): void {
    this.telegramBot = new TelegramBot(token, {polling: true});
    this.telegramBot.onText(regex, async (msg: TelegramBot.Message) => {
      const chatId = msg.chat.id;
      if (this.chatroomIDs.contains(chatId)) {
        this.telegramBot.sendMessage(chatId, "The notification service is running").catch(console.error);
      } else {
        this.chatroomIDs.push(chatId);
        this.telegramBot.sendMessage(chatId, "started the notification service").catch(console.error);
      }
    });
  }

  public newPullRequest(notification: Notification): Promise<void> {
    // TODO
    return undefined;
  }

  public updatedPullRequest(notification: Notification): Promise<void> {
    // TODO
    return undefined;
  }

  public revisionNeeded(notification: Notification): Promise<void> {
    // TODO
    return undefined;
  }

  public autoMerging(notification: Notification): Promise<void> {
    // TODO
    return undefined;
  }
}
