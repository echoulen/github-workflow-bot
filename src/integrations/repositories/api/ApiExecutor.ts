import {List} from "immutable";

export interface ApiExecutor {
  getCurrentUser<User>(): Promise<User>;
  addComment<Comment>(project, pr, comment: string): Promise<Comment>;
  getComments<Comment>(project, pr): Promise<List<Comment>>;
  setLabel(project, pr, labels: string): Promise<void>;
  merge(project, pr): Promise<void>;
}
