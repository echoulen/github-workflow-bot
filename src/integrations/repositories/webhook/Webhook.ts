import {Notification} from "@type/notification/Notification";

export interface Webhook {
  isPullRequest(hook): boolean;
  isComment(hook): boolean;
  processPullRequest(hook): Promise<Notification>;
  processComment(hook): Promise<Notification>;
}
