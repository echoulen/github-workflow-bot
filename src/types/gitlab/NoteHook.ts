export interface NoteHook {
    object_kind: "note";
    user: {
        name: string;
        username: string;
        avatar_url: string;
    };
    project_id: number;
    project: {
        name: string;
        description: string;
        web_url: string;
        avatar_url: string;
        git_ssh_url: string;
        git_http_url: string;
        namespace: string;
        visibility_level: number;
        path_with_namespace: string;
        default_branch: string;
        ci_config_path: string;
        homepage: string;
        url: string;
        ssh_url: string;
        http_url: string;
    };
    object_attributes: {
        id: number;
        note: string;
        noteable_type: "MergeRequest" | any;
        author_id: number;
        created_at: string;
        updated_at: string;
        project_id: number;
        attachment: any;
        line_code: string;
        commit_id: string;
        noteable_id: number;
        system: boolean;
        st_diff: null;
        updated_by_id: number;
        type: "DiscussionNote" | "DiffNote" | null;
        position: string;
        original_position: string;
        resolved_at: string;
        resolved_by_id: number;
        discussion_id: string;
        change_position: string;
        resolved_by_push: string;
        url: string;
    };
    repository: {
        name: string;
        url: string;
        description: string;
        homepage: string;
    };
    "merge_request": {
        id: number;
        target_branch: string;
        source_branch: string;
        source_project_id: number;
        author_id: number;
        assignee_id: number;
        title: string;
        created_at: string;
        updated_at: string;
        milestone_id: string;
        state: "opened" | "closed",
        merge_status: "can_be_merged" | "cannot_be_merged",
        target_project_id: number;
        iid: number;
        description: string;
        updated_by_id: string;
        merge_error: any;
        merge_params: {
            force_remove_source_branch: string;
        };
        merge_when_pipeline_succeeds: boolean;
        merge_user_id: number;
        merge_commit_sha: string;
        deleted_at: string;
        in_progress_merge_commit_sha: string;
        lock_version: string;
        time_estimate: number;
        last_edited_at: string;
        last_edited_by_id: number;
        head_pipeline_id: number;
        ref_fetched: boolean;
        merge_jid: string;
        source: {
            name: string;
            description: string;
            web_url: string;
            avatar_url: null,
            git_ssh_url: string;
            git_http_url: string;
            namespace: string;
            visibility_level: number;
            path_with_namespace: string;
            default_branch: string;
            ci_config_path: string;
            homepage: string;
            url: string;
            ssh_url: string;
            http_url: string;
        };
        target: {
            name: string;
            description: string;
            web_url: string;
            avatar_url: string;
            git_ssh_url: string;
            git_http_url: string;
            namespace: string;
            visibility_level: number;
            path_with_namespace: string;
            default_branch: string;
            ci_config_path: string;
            homepage: string;
            url: string;
            ssh_url: string;
            http_url: string;
        };
        last_commit: {
            id: string;
            message: string;
            timestamp: string;
            url: string;
            author: {
                name: string;
                email: string;
            };
        };
        work_in_progress: boolean;
        total_time_spent: number;
        human_total_time_spent: string;
        human_time_estimate: string;
    };
}
