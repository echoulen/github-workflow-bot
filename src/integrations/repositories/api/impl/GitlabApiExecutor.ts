import {ConfigManager} from "@config/ConfigManager";
import {ApiExecutor} from "@repositories/api/ApiExecutor";
import axios from "axios";
import {List} from "immutable";
import {injectable} from "inversify";
import "reflect-metadata";

@injectable()
export class GitlabApiExecutor implements ApiExecutor {

  private readonly url = ConfigManager.getConfig().repository.url;
  private readonly headers = {"PRIVATE-TOKEN": ConfigManager.getConfig().repository.token};

  public async getBotUser<User>(): Promise<User> {
    const res = await axios.get(`${this.url}/api/v4/user`, {headers: this.headers});
    return res.data;
  }

  public async getPullRequest<PullRequest>(project: number, pr: number): Promise<PullRequest> {
    const res = await axios.get(`${this.url}/api/v4/projects/${project}/merge_requests/${pr}`, {headers: this.headers});
    return res.data;
  }

  public async addComment<T>(project: number, pr: number, comment: string): Promise<T> {
    const res = await axios.post(
      `${this.url}/api/v4/projects/${project}/merge_requests/${pr}/notes`,
      {body: comment},
      {headers: this.headers},
    );
    return res.data;
  }

  public async getComments<T>(project: number, pr: number): Promise<List<T>> {
    const res = await axios.get(`${this.url}/api/v4/projects/${project}/merge_requests/${pr}/notes`, {headers: this.headers});
    return List<T>(res.data);
  }

  public async setLabel(project: number, pr: number, labels: string): Promise<void> {
    await axios.put(
      `${this.url}/api/v4/projects/${project}/merge_requests/${pr}`,
      {labels},
      {headers: this.headers},
    );
  }

  public async merge(project: number, pr: number): Promise<void> {
    await axios.put(`${this.url}/api/v4/projects/${project}/merge_requests/${pr}/merge`, {}, {headers: this.headers});
  }
}
