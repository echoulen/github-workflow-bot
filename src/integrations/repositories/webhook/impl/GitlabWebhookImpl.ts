import {ConfigManager} from "@config/ConfigManager";
import {BeanTypes} from "@constants/BeanType";
import {Label} from "@constants/Label";
import {Message} from "@constants/Message";
import {ApiExecutor} from "@repositories/api/ApiExecutor";
import {updateLabels} from "@repositories/utils/updateLabels";
import {Webhook} from "@repositories/webhook/Webhook";
import {Comment} from "@type/gitlab/Comment";
import {MergeRequest} from "@type/gitlab/MergeRequest";
import {MergeRequestHook} from "@type/gitlab/MergeRequestHook";
import {NoteHook} from "@type/gitlab/NoteHook";
import {User} from "@type/gitlab/User";
import {List} from "immutable";
import {inject, injectable} from "inversify";
import "reflect-metadata";
import {sprintf} from "sprintf-js";

@injectable()
export class GitlabWebhookImpl implements Webhook {

  @inject(BeanTypes.API_EXECUTOR)
  private apiExecutor: ApiExecutor;
  private config = ConfigManager.getConfig();
  private autoMergeSec = 25;

  public isPullRequest(hook: MergeRequestHook): boolean {
    return hook.object_kind === "merge_request";
  }

  public isComment(hook: NoteHook): boolean {
    return hook.object_kind === "note";
  }

  public async processPullRequest(hook: MergeRequestHook): Promise<void> {
    const projectId = hook.object_attributes.source_project_id;
    const mergeRequestId = hook.object_attributes.iid;
    const triggerBranch = hook.object_attributes.target_branch;
    const labels: List<string> = List(hook.labels).map<string>((it) => it.title);
    const defaultBranch = hook.project.default_branch;
    const bot = await this.apiExecutor.getBotUser<User>();
    const mergeRequest = await this.apiExecutor.getPullRequest<MergeRequest>(projectId, mergeRequestId);

    if (!mergeRequestId || triggerBranch !== defaultBranch) {
      return;
    }

    if (hook.object_attributes.action === "open") {
      const username = hook.user.username;
      await this.apiExecutor.addComment(
        projectId,
        mergeRequestId,
        sprintf(Message.AWAIT_REVIEWER, username),
      );
      const newLabels = updateLabels(labels, [Label.AWAIT_REVIEWER], []);
      await this.apiExecutor.setLabel(projectId, mergeRequestId, newLabels);
    } else if (
      hook.object_attributes.action === "update" &&
      hook.user.username !== bot.username &&
      mergeRequest.author.username === hook.user.username &&
      List(labels).contains(Label.REVISION_NEEDED)) {

      const newLabels = updateLabels(labels, [Label.UPDATED], [Label.REVISION_NEEDED]);
      await this.apiExecutor.setLabel(projectId, mergeRequestId, newLabels);
    }
  }

  public async processComment(hook: NoteHook): Promise<void> {
    const bot = await this.apiExecutor.getBotUser<User>();

    if (hook.object_attributes.noteable_type !== "MergeRequest") {
      return;
    }

    if (hook.user.username === bot.username) {
      return;
    }

    const projectId = hook.merge_request.target_project_id;
    const mergeRequestId = hook.merge_request.iid;
    const mergeRequest = await this.apiExecutor.getPullRequest<MergeRequest>(projectId, mergeRequestId);
    const labels = List<string>(mergeRequest.labels);

    if (hook.merge_request.work_in_progress || hook.merge_request.merge_status !== "can_be_merged") {
      return;
    }

    if (hook.object_attributes.type === "DiscussionNote" || hook.object_attributes.type === "DiffNote") {
      await this.apiExecutor.setLabel(
        projectId,
        mergeRequestId,
        updateLabels(labels, [Label.REVISION_NEEDED], [Label.AWAIT_REVIEWER, Label.UPDATED, Label.AWAIT_AUTO_MERGE, Label.ONE_LGTM]),
      );
    } else if (hook.object_attributes.note.toUpperCase().includes("LGTM")) {

      const comments = await this.apiExecutor.getComments<Comment>(projectId, mergeRequestId);
      const lgtmReviewers = comments.filter((it) => it.body.toUpperCase().includes("LGTM")).map((comment) => comment.author.name).toSet();
      const currentLgtmCount = lgtmReviewers.count();
      const awaitAutoMergeMessage = sprintf(Message.AWAIT_AUTO_MERGE, this.autoMergeSec);
      const triggerMessage = comments.find((it) => it.body === awaitAutoMergeMessage);

      if (currentLgtmCount >= this.config.lgtm.count && !triggerMessage) {
        await this.apiExecutor.addComment(projectId, mergeRequestId, awaitAutoMergeMessage);
        const cleanLabels = updateLabels(
          labels,
          [Label.AWAIT_AUTO_MERGE],
          [Label.AWAIT_REVIEWER, Label.ONE_LGTM, Label.REVISION_NEEDED, Label.UPDATED],
        );
        await this.apiExecutor.setLabel(projectId, mergeRequestId, cleanLabels);

        setTimeout(async () => {
          await this.apiExecutor.merge(projectId, mergeRequestId);
          const finalMergeRequest = await this.apiExecutor.getPullRequest<MergeRequest>(projectId, mergeRequestId);
          const finalLabels = List<string>(finalMergeRequest.labels);
          const finalNewLabels = updateLabels(finalLabels, [], [Label.AWAIT_AUTO_MERGE, Label.ONE_LGTM, Label.UPDATED]);
          await this.apiExecutor.setLabel(projectId, mergeRequestId, finalNewLabels);
        }, this.autoMergeSec * 1000);
      } else {
        const newLabels = updateLabels(labels, [Label.AWAIT_REVIEWER, Label.ONE_LGTM], [Label.REVISION_NEEDED]);
        await this.apiExecutor.setLabel(projectId, mergeRequestId, newLabels);
      }
    }
  }
}
