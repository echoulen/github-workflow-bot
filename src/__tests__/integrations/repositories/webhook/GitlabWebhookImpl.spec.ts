import {ConfigManager} from "@config/ConfigManager";
import {BeanTypes} from "@constants/BeanType";
import {ApiExecutor} from "@repositories/api/ApiExecutor";
import {GitlabApiExecutor} from "@repositories/api/impl/GitlabApiExecutor";
import {GitlabWebhookImpl} from "@repositories/webhook/impl/GitlabWebhookImpl";
import {Webhook} from "@repositories/webhook/Webhook";
import {MockMergeRequestHook} from "@tests/test-helper/mock-data/gitlab/hook/MockMergeRequestHook";
import {MockNoteHook} from "@tests/test-helper/mock-data/gitlab/hook/MockNoteHook";
import {MockComments} from "@tests/test-helper/mock-data/gitlab/MockComments";
import {MockLGTMComments} from "@tests/test-helper/mock-data/gitlab/MockLGTMComments";
import {MockMergeRequest} from "@tests/test-helper/mock-data/gitlab/MockMergeRequest";
import {NoteHook} from "@type/gitlab/NoteHook";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import {Container} from "inversify";
import sinon from "sinon";

describe("GitlabWebhookImpl spec", () => {
  jest.useFakeTimers();
  const url = ConfigManager.getConfig().repository.url;
  const project = 123;
  const pr = 54088;
  const mock = new MockAdapter(axios);
  let appContainer: Container;
  let webhook: Webhook;

  beforeAll(() => {
    appContainer = new Container();
    appContainer.bind<Webhook>(BeanTypes.WEBHOOK).to(GitlabWebhookImpl).inSingletonScope();
    appContainer.bind<ApiExecutor>(BeanTypes.API_EXECUTOR).to(GitlabApiExecutor).inSingletonScope();
    webhook = appContainer.get<Webhook>(BeanTypes.WEBHOOK);

    mock.onGet(`${url}/api/v4/user`).reply(200, {id: 123, name: "bot-user", username: "bot"});
    mock.onGet(`${url}/api/v4/projects/${project}/merge_requests/${pr}`).reply(200, MockMergeRequest);
    mock.onGet(`${url}/api/v4/projects/${project}/merge_requests/${pr}/notes`).reply(200, MockComments);

    mock.onGet(`${url}/api/v4/projects/23/merge_requests/1267`).reply(200, MockMergeRequest);
    mock.onGet(`${url}/api/v4/projects/23/merge_requests/1267/notes`).reply(200, MockLGTMComments);
    mock.onPut(`${url}/api/v4/projects/23/merge_requests/1267`).reply(200, {});
  });

  it("should isPullRequest proper", () => {
    const result = webhook.isPullRequest(MockMergeRequestHook);
    expect(result).toBe(true);

    const result2 = webhook.isPullRequest(MockNoteHook);
    expect(result2).toBe(false);
  });

  it("should isComment proper", () => {
    const result = webhook.isComment(MockMergeRequestHook);
    expect(result).toBe(false);

    const result2 = webhook.isComment(MockNoteHook);
    expect(result2).toBe(true);
  });

  it("should processPullRequest proper", async () => {
    const mockMethod = sinon.mock(GitlabApiExecutor.prototype);
    mockMethod.expects("setLabel").once().returns(new Promise((resolve) => resolve()));
    const notify = await webhook.processPullRequest(MockMergeRequestHook);
    expect(notify).toMatchSnapshot();
    mockMethod.verify();
    mockMethod.restore();
  });

  it("should processComment proper with DiscussionNote", async () => {
    const mockMethod = sinon.mock(GitlabApiExecutor.prototype);
    mockMethod.expects("setLabel").once().returns(new Promise((resolve) => resolve()));
    mockMethod.expects("merge").never().returns(new Promise((resolve) => resolve()));
    const notify = await webhook.processComment(MockNoteHook);
    expect(notify).toMatchSnapshot();
    mockMethod.verify();
    mockMethod.restore();
  });

  it("should processComment proper with LGTM", async () => {
    const mockMethod = sinon.mock(GitlabApiExecutor.prototype);
    mockMethod.expects("setLabel").once().returns(new Promise((resolve) => resolve()));
    mockMethod.expects("merge").never().returns(new Promise((resolve) => resolve()));
    const notify = await webhook.processComment(MockNoteHook);
    expect(notify).toMatchSnapshot();
    mockMethod.verify();
    mockMethod.restore();
  });

  it("should processComment proper with LGTM and auto merge", async () => {
    const mockMethod = sinon.mock(GitlabApiExecutor.prototype);
    mockMethod.expects("setLabel").once().returns(new Promise((resolve) => resolve()));
    mockMethod.expects("addComment").once().returns(new Promise((resolve) => resolve()));
    mockMethod.expects("merge").once().returns(new Promise((resolve) => resolve()));

    const lgtmNoteHook: NoteHook = {
      ...MockNoteHook,
      object_attributes: {
        ...MockNoteHook.object_attributes,
        note: "lgtm !!",
        type: "note",
      },
    };
    const notify = await webhook.processComment(lgtmNoteHook);
    expect(notify).toMatchSnapshot();
    jest.advanceTimersByTime(25000);
    mockMethod.verify();
    mockMethod.restore();
  });

  afterAll(() => {
    appContainer.restore();
    mock.reset();
    mock.restore();
  });
});
