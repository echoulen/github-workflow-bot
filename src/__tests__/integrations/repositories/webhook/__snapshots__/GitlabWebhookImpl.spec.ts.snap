// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitlabWebhookImpl spec should processComment proper with DiscussionNote 1`] = `
Object {
  "avatarUrl": "http://gitlab.com/uploads/-/system/user/avatar/82/avatar.png",
  "description": "test description",
  "link": "http://gitlab.com/LT/V3Desktop/merge_requests/1333",
  "title": "test title",
  "type": "NotificationType.REVISION_NEEDED",
  "username": "merge.bot",
}
`;

exports[`GitlabWebhookImpl spec should processComment proper with LGTM 1`] = `
Object {
  "avatarUrl": "http://gitlab.com/uploads/-/system/user/avatar/82/avatar.png",
  "description": "test description",
  "link": "http://gitlab.com/LT/V3Desktop/merge_requests/1333",
  "title": "test title",
  "type": "NotificationType.REVISION_NEEDED",
  "username": "merge.bot",
}
`;

exports[`GitlabWebhookImpl spec should processComment proper with LGTM and auto merge 1`] = `
Object {
  "avatarUrl": "http://gitlab.com/uploads/-/system/user/avatar/82/avatar.png",
  "description": "test description",
  "link": "http://gitlab.com/LT/V3Desktop/merge_requests/1333",
  "title": "test title",
  "type": "NotificationType.AWAIT_AUTO_MERGE",
  "username": "merge.bot",
}
`;

exports[`GitlabWebhookImpl spec should processPullRequest proper 1`] = `
Object {
  "avatarUrl": "http://gitlab.com/uploads/-/system/user/avatar/82/avatar.png",
  "description": "test description",
  "link": "http://gitlab.com/LT/V3Desktop/merge_requests/1333",
  "title": "test title",
  "type": "NotificationType.UPDATE",
  "username": "test user",
}
`;
